scio-test/src/main/scala/com/spotify/scio/testing/JobTest.scala

```diff
@@ -98,11 +100,24 @@ object JobTest {
     }
 
     /**
-     * Feed an input to the pipeline being tested. Note that `TestIO[T]` must match the one used
-     * inside the pipeline, e.g. `AvroIO[MyRecord]("in.avro")` with
-     * `sc.avroFile[MyRecord]("in.avro")`.
+     * Feed an input in the form of a raw Iterable[T] to the pipeline being tested. Note that
+     * `TestIO[T]` must match the one used inside the pipeline, e.g. `AvroIO[MyRecord]("in.avro")`
+     * with `sc.avroFile[MyRecord]("in.avro")`.
      */
-    def input[T](io: ScioIO[T], value: Iterable[T]): Builder = {
+    def input[T](io: ScioIO[T], value: Iterable[T]): Builder =
+      input(io, IterableInputSource(value))
+
+    /**
+     * Feed an input in the form of a PTransform[PBegin, PCollection[T] to the pipeline being
+     * tested. Note that PTransform inputs may not be supported for all TestIO[T] types.
+     */
+    def pInput[T](
```

Comment:
What's the reason you can't overload here?
What about `def input[T](io)(transform)`? Also if you make `transform` a `fn: TestStream => TestStream`, user code could look like this:
```scala
.input(TextIO[String]) { _.addElements("a b b c").... }
```

WDYT?

Reply:
The reason is that the Scala compiler can't infer which method to use when the generic type is not known:

```
def input[T](io: ScioIO[T], value: Iterable[T]): Builder = ...
def input[T](io: ScioIO[T], transform: PTransform[_ >: PBegin <: PInput, PCollection[T]]): Builder = ...

final case class ReadIO[T](id: String) extends TestIO[T]

JobTest[ReadAllJob.type]
  .input(ReadIO("a"), Seq("a1", "a2")) // fails, scala cannot resolve overloaded method

JobTest[ReadAllJob.type]
  .input(ReadIO[String]("a"), Seq("a1", "a2")) // succeeds
```

I don't know of a way around this without changing the original signature of `def input`, or requiring users to be explicit about `[T]` every time they use `ScioIO`s :|

Reply:
for the second point - you mean replacing the signature ` PTransform[PBegin, ...` with `TestStream` specifically--or adding a new method for `TestStream` type? IMO the signature `.input(TextIO[String]) { _.addElements("a b b c").... }` reads a little unclear, maybe `.inputStream(TextIO[String]) { _.addElements("a b b c").... }` ?

Reply:
:+1: on `.inputStream`. In that case I guess we can just make the 2nd arg `TestStream`, and it doesn't matter whether it's curried or not.

Reply:
cool, i'll just make this specific to TestStream, then! I think currying could have been a little awkward since `.addElements(...)` returns a `TestStream.Builder`, which only gets converted to `TestStream` when `advanceWatermarkToInfinity()` is called on it.